python/pyspark/sql/connect/udf.py

```diff
@@ -0,0 +1,176 @@
+"""
+import functools
+from typing import Callable, Any, TYPE_CHECKING, Optional
+
+from pyspark import cloudpickle
+from pyspark.sql.connect.expressions import (
+    ColumnReference,
+    PythonUDF,
+    ScalarInlineUserDefinedFunction,
+)
+from pyspark.sql.connect.column import Column
+from pyspark.sql.types import DataType, StringType
+
+
+if TYPE_CHECKING:
+    from pyspark.sql.connect._typing import (
+        ColumnOrName,
+        DataTypeOrString,
+        UserDefinedFunctionLike,
+    )
+    from pyspark.sql.types import StringType
+
+
+def _create_udf(
+    f: Callable[..., Any],
+    returnType: "DataTypeOrString",
+    evalType: int,
+    name: Optional[str] = None,
+    deterministic: bool = True,
+) -> "UserDefinedFunctionLike":
+    # Set the name of the UserDefinedFunction object to be the name of function f
+    udf_obj = UserDefinedFunction(
+        f, returnType=returnType, name=name, evalType=evalType, deterministic=deterministic
+    )
+    return udf_obj._wrapped()
+
+
+class UserDefinedFunction:
+    """
+    User defined function in Python
+
+    Notes
+    -----
+    The constructor of this class is not supposed to be directly called.
+    Use :meth:`pyspark.sql.functions.udf` or :meth:`pyspark.sql.functions.pandas_udf`
+    to create this instance.
+    """
+
+    def __init__(
+        self,
+        func: Callable[..., Any],
+        returnType: "DataTypeOrString" = StringType(),
+        name: Optional[str] = None,
+        evalType: int = 100,
+        deterministic: bool = True,
+    ):
+        if not callable(func):
+            raise TypeError(
+                "Invalid function: not a function or callable (__call__ is not defined): "
+                "{0}".format(type(func))
+            )
+
+        if not isinstance(returnType, (DataType, str)):
+            raise TypeError(
+                "Invalid return type: returnType should be DataType or str "
+                "but is {}".format(returnType)
+            )
+
+        if not isinstance(evalType, int):
+            raise TypeError(
+                "Invalid evaluation type: evalType should be an int but is {}".format(evalType)
+            )
+
+        self.func = func
+        self._returnType = returnType
+        self._name = name or (
+            func.__name__ if hasattr(func, "__name__") else func.__class__.__name__
+        )
+        self.evalType = evalType
+        self.deterministic = deterministic
+
+    def __call__(self, *cols: "ColumnOrName") -> Column:
+        arg_cols = [
+            col if isinstance(col, Column) else Column(ColumnReference(col)) for col in cols
+        ]
+        arg_exprs = [col._expr for col in arg_cols]
+        data_type_str = (
+            self._returnType.json() if isinstance(self._returnType, DataType) else self._returnType
+        )
+        py_udf = PythonUDF(
+            output_type=data_type_str,
+            eval_type=self.evalType,
+            command=cloudpickle.dumps(self.func),
```

Comment:
This line has a bug unfortunately. The serialization protocol is that it always serializes a pair of `(function, returntype)`. Once this is fixed the code works for the UDFs.

Reply:
Fixed! Would you mind sharing where the serialization protocol is defined?

Reply:
Found it, serializing `returntype` may introduce a bug

```py
>>> @udf('string')
... def f(x):
...   return str(x)
... 
>>> spark.range(2).withColumn('to_str', f('id')).show()
...
AttributeError: 'str' object has no attribute 'needConversion'
```
Let me fix it.